### PR TITLE
[InstSimplify] Simplify `(fmul -x, +/-0)` -> `-/+0`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstSimply] Add tests for simplify `(fmul -x, +/-0)`; NFC**
- **[InstSimply] Simplify `(fmul -x, +/-0)` -> `-/+0`**

Proofs: https://alive2.llvm.org/ce/z/WUSvmV

### CODE REVIEW

File: llvm/test/Transforms/InstSimplify/floating-point-arithmetic.ll [hunk 1]
```diff
@@ -211,6 +211,22 @@ define double @fmul_nnan_ninf_nneg_n0.0_commute(i127 %x) {
...
+}
+
+define float @src_mul_zero_neg(float nofpclass(inf nan pzero psub pnorm) %f) {
+; CHECK-LABEL: @src_mul_zero_neg(
+; CHECK-NEXT:    ret float -0.000000e+00
+;
+  %r = fmul float %f, 0.0
+  ret float %r
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstSimplify/floating-point-arithmetic.ll:229]: Can you also test the FMA / fmuladd cases? Also vectors

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 2]
```diff
@@ -5767,11 +5767,18 @@ static Value *simplifyFMAFMul(Value *Op0, Value *Op1, FastMathFlags FMF,
...
-      return Op1;
+    if (Known.isKnownNever(fcInf | fcNan)) {
+      // +normal number * (-)0.0 --> (-)0.0
+      if (Known.SignBit == false)
+        return Op1;
+      // -normal number * (-)0.0 --> -(-)0.0
+      if (Known.SignBit == true)
+        return match(Op1, m_PosZeroFP())
+                   ? ConstantFP::getNegativeZero(Op0->getType())
+                   : ConstantFP::getZero(Op0->getType());
```
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:5780]: Can also just pass the bool condition to the parameter to getZero to make it signed

File: llvm/test/Transforms/InstSimplify/floating-point-arithmetic.ll (hunk 1 above)
Comment (goldsteinn) [reply llvm/test/Transforms/InstSimplify/floating-point-arithmetic.ll:229]: Done, although neither of those cases get simplified by `instsimplify` alone.

File: llvm/lib/Analysis/InstructionSimplify.cpp (hunk 2 above)
Comment (goldsteinn) [reply llvm/lib/Analysis/InstructionSimplify.cpp]: Switching to just `fneg`, that's necessary for handling mismatch vector.

File: llvm/test/Transforms/InstSimplify/floating-point-arithmetic.ll [hunk 3]
```diff
@@ -211,6 +211,50 @@ define double @fmul_nnan_ninf_nneg_n0.0_commute(i127 %x) {
...
+  ret <2 x float> %r
+}
+
+
+define float @src_muladd_zero_neg(float nofpclass(inf nan pzero psub pnorm) %f, float %add) {
+; CHECK-LABEL: @src_muladd_zero_neg(
+; CHECK-NEXT:    [[R:%.*]] = call float @llvm.fmuladd.f32(float [[F:%.*]], float 0.000000e+00, float [[ADD:%.*]])
+; CHECK-NEXT:    ret float [[R]]
+;
+  %r = call float @llvm.fmuladd.f32(float %f, float 0.0, float %add)
```
Comment (arsenm) [inline llvm/test/Transforms/InstSimplify/floating-point-arithmetic.ll:244]: I think the case fmuladd/fma will be handled is if the last operand is 0

Comment (goldsteinn) [reply llvm/test/Transforms/InstSimplify/floating-point-arithmetic.ll:244]: Doesn't seem to be the case, although note `instcombine` does handle the `fmuladd`/`fma`. Just not simplify.

**Review by arsenm — APPROVED** [review]

**Pull request closed without merging**